sourcegraph/cody
# fix/gql: Cache the SiteProductVersion query for up to 10 minutes

We hammer the SiteProductVersion GraphQL endpoint really hard. This adds a cache in front of it to reduce the heavy traffic to that endpoint. The cache also makes the product faster.

Changes the `currentSiteVersion` to use this cache directly instead of burying it in an observable where it may cache transient errors like network errors.

I tried writing an exponential-backoff-and-retry Observable, and it is possible, but it is a bit tortured because the downstream consumer pings the upstream "(re)try" producer to cause the Observable to re-emit. I judged caching in the GraphQL client and retrying when called was simpler.

One subtlety among many: Aborts cause exponential backoff. This is a feature: it prevents a misbehaving client (which we have suffered recently) from initiating and then immediately aborting many requests rapidly.

## Test plan

New automated test:

```
pnpm -C lib/shared unit -- cache.test.ts
```

Manual test:

- Verify spans like `graphql.fetch.SiteProductVersion.cached - 0.473875ms` appear. For example, run the product in the debugger and check the Debug Console for those entries.
- Change accounts, and verify cache miss spans like `graphql.fetch.SiteProductVersion - 261.213334ms` appear.

## Changelog

- SiteProductVersion, a heavily used GraphQL query, is cached for up to 10 minutes. Failures are retried with exponential backoff.
